src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -481,11 +477,52 @@ EXTERN_C intptr_t RhGetCurrentThunkContext()
 //  thread        - thread to attach
 extern "C" void PalAttachThread(void* thread)
 {
-#if HAVE_THREAD_LOCAL
     tls_destructionMonitor.SetThread(thread);
-#else
-    __cxa_thread_atexit(RuntimeThreadShutdown, thread, &__dso_handle);
-#endif
+}
+
+struct Link
+{
+    Link* Next;
+    Destructor Dtor;
+    void* Arg;
+};
+
+static pthread_key_t key;
+static pthread_once_t once = PTHREAD_ONCE_INIT;
+
+static void CallDestructors(void* obj)
+{
+    Link* next;
+    Link* link = (Link*)obj;
+    while (link)
+    {
+        next = link->Next;
+        link->Dtor(link->Arg);
+        free(link);
+        link = next;
+    }
+}
+
+static void CreateKey(void)
+{
+    pthread_key_create(&key, CallDestructors);
```

Comment:
This can be done as part of `PalInit()`

Reply:
Moved `pthread_once` call to `PalInit` (which takes this factory as an arg). It is a bit safer with `PTHREAD_ONCE_INIT`. If we are sure that `PalInit` can not run into race conditions, I can inline it and remove `PTHREAD_ONCE_INIT`.